pylint/pyreverse/writer.py

```diff
@@ -129,6 +179,12 @@ def write_classes(self, diagram: ClassDiagram) -> None:
             )
         # inheritance links
         for rel in diagram.get_relationships("specialization"):
```

Comment:
```suggestion
        for rel in diagram.get_relationships("specialization", depth=self.max_depth):
```
Wonder if it's possible of filtering on depth directly in ``get_relationships`` instead. Seems like it could reduce duplication of ``.should_show_node`` checks when looping on it

Reply:
Thanks for the suggestion! I would agree, that handling everything regarding filtering in one place would benefit maintainability. However I think we still need node filtering when emitting nodes, to determine which ones to show in the final diagram. Not sure about splitting this up into 2 files either. What do you think?

Reply:
Actually, I feel like we could simplify this, by introducing `get_nodes()` alongside `get_relationships()`. Then the writer would look something like this.
```python
def write_packages(self, diagram: PackageDiagram) -> None:
    for module in diagram.get_nodes(depth=self.max_depth):
        # Emit node logic
    for rel in diagram.get_relationships(depth=self.max_depth):
        # Emit edge logic
```
That way all the filtering logic would be handled consistently inside the diagram classes and the printer would only need to focus on rendering the final components. What do you think?

Reply:
Looks great !

Reply:
Since this is a big change, I am not sure how to approach this. Are you ok with doing this in this PR?